### PR TITLE
feat(core): allow to add precompiled tokens via a provider

### PR DESCRIPTION
**Pull request by tbosch**

### CODE REVIEW

File: modules/@angular/compiler-cli/integrationtest/src/module_fixtures.ts [hunk 1]
```diff
@@ -53,3 +53,17 @@ export class SomeModule {
 })
 export class SomeModuleUsingParentComp {
 }
+
+export const SOME_TOKEN = new OpaqueToken('someToken');
+
+export function provideValueWithPrecompile(value: any) {
+  return [
+    {provide: SOME_TOKEN, useValue: value},
+    {provide: ANALYZE_FOR_PRECOMPILE, useValue: value},
```
Comment (vsavkin) [inline modules/@angular/compiler-cli/integrationtest/src/module_fixtures.ts]: should it be multi? I know it won't affect anything, but semantically it acts as multi as we do not override previously seen ANALYZE_FOR_PRECOMPILE

File: modules/@angular/compiler-cli/integrationtest/test/app_module_spec.ts [hunk 2]
```diff
@@ -26,6 +26,15 @@ describe('AppModule', () => {
     expect(compRef.instance instanceof SomeComp).toBe(true);
   });
 
+  it('should support precompile via the ANALYZE_FOR_PRECOMPILE provider and function providers in components',
+     () => {
+       var moduleRef = createModule(SomeModuleWithAnalyzePrecompileProviderNgFactory);
```
Comment (vsavkin) [inline modules/@angular/compiler-cli/integrationtest/test/app_module_spec.ts]: can you make these const?

File: modules/@angular/compiler/src/metadata_resolver.ts [hunk 3]
```diff
@@ -410,23 +412,51 @@ export class CompileMetadataResolver {
...
+        compileProviders.push(compileProvider);
+      }
+    });
+    return compileProviders;
+  }
+
+  getPrecompileComponentsFromProvider(provider: Provider): cpl.CompileTypeMetadata[] {
+    let components: cpl.CompileTypeMetadata[] = [];
+    let collectedIdentifiers: cpl.CompileIdentifierMetadata[] = [];
+    if (provider.useFactory || provider.useExisting || provider.useClass) {
```
Comment (vsavkin) [inline modules/@angular/compiler/src/metadata_resolver.ts]: maybe add a check for multi here

File: modules/@angular/compiler-cli/integrationtest/src/module_fixtures.ts (hunk 1 above)
Comment (tbosch) [inline modules/@angular/compiler-cli/integrationtest/src/module_fixtures.ts]: changed

File: modules/@angular/compiler-cli/integrationtest/test/app_module_spec.ts (hunk 2 above)
Comment (tbosch) [inline modules/@angular/compiler-cli/integrationtest/test/app_module_spec.ts]: changed

File: modules/@angular/compiler/src/metadata_resolver.ts (hunk 3 above)
Comment (tbosch) [inline modules/@angular/compiler/src/metadata_resolver.ts]: done

**Pull request closed without merging**